File: Codigos estudiantes por lenguaje/C++/Cristian Maldonado/Bitwise/Code_Conversion_Gray_to_Binary_And_Reverse.cpp
#include <iostream>
//--------------//
#define FAST ios_base::sync_with_stdio(false);cout.tie(NULL);

using namespace std;

char xor_c(char a, char b){
    return (a == b)? '0':'1';
}

string binaryToGray(string binary){
    string gray = "";
    int tam = binary.size();
    gray += binary[0];
    for(int i = 1; i < tam; i++){
        gray += xor_c(binary[i - 1], binary[i]);
    }
    return gray;
}

char flip(char c){
    return (c == '0') ? '1': '0';
}

string grayToBinary(string gray){
    string binary = "";
    int tam = gray.size();
    binary += gray[0];
    for(int i = 1; i < tam; i++){
        if(gray[i] == '0'){
            binary += binary[i - 1];
        }else{
            binary += flip(binary[i - 1]);
        }
    }
    return binary;
}

int main()
{
    FAST
    string b1 = "111010";
    string res1 = binaryToGray(b1);
    string res2 = grayToBinary(res1);
    cout<<"De binario a Gray: "<<res1<<endl;
    cout<<"De Gray a Binario: "<<res2<<endl;
}

File: Codigos estudiantes por lenguaje/Java/Cristian Maldonado/Bitwise/Code_Conversion_Gray_to_Binary_And_Reverse.java
public class Code_Conversion_Gray_to_Binary_And_Reverse {

    static String binaryToGray(String binary) {
        String gray = "";
        gray += binary.charAt(0);
        for (int i = 1; i < binary.length(); i++) {
            gray += xor_c(binary.charAt(i - 1), binary.charAt(i));
        }
        return gray;
    }

    static char xor_c(char a, char b) {
        return a == b ? '0' : '1';
    }

    static String grayToBinary(String gray) {
        String binary = "";
        binary += gray.charAt(0);
        for (int i = 1; i < gray.length(); i++) {
            if (gray.charAt(i) == '0') {
                binary += binary.charAt(i - 1);
            } else {
                binary += flip(binary.charAt(i - 1));
            }
        }
        return binary;
    }

    static char flip(char c) {
        return c == '0' ? '1' : '0';
    }
    
    public static void main(String[] args) {
        String b1 = "01001";
        String res = binaryToGray(b1);
        System.out.println("Binario a Gray: "+res);
        String res2 = grayToBinary(res);
        System.out.println("Gray a Binario: "+res2);
    }   
}
